linmod/eval.py

```diff
@@ -1,124 +1,210 @@
+import numpy as np
 import polars as pl
 
 from linmod.utils import pl_list_cycle, pl_norm
 
 
-def _merge_samples_and_data(samples, data):
-    r"""
-    Join the forecast samples and raw data dataframes, assuming they have the
-    standard format.
-
-    Also compute the true proportions from the raw data.
-    """
-    result = (
-        data.with_columns(
-            phi=(
-                pl.col("count") / pl.sum("count").over("fd_offset", "division")
-            ),
-        )
-        .drop("count")
-        .join(
-            samples,
-            on=("fd_offset", "division", "lineage"),
-            how="left",
-            suffix="_sampled",
+def multinomial_count_sampler(N, phi, rng):
+    return rng.multinomial(N, phi)
+
+
+class ProportionsEvaluator:
+    def __init__(self, samples: pl.LazyFrame, data: pl.LazyFrame):
+        """
+        `samples` should have the standard model output format.
+        `data` should have the standard model input format.
+        """
+
+        # Join the forecast samples and raw data dataframes.
+        # Also compute the true proportions from the raw data.
+        self.df = (
+            data.with_columns(
+                phi=(
+                    pl.col("count")
+                    / pl.sum("count").over("fd_offset", "division")
+                ),
+            )
+            .drop("count")
+            # NaN will appear on division-days where none of the lineages
+            # were sampled. We will filter these out.
+            # TODO: Is this good to do?
```

Comment:
I think this is a signal that you want a separate function for computing phi from counts, that can have branching logic for what to do with a denominator count of 0

so that this becomes `self.df = data.with_columns(phi=pl.col('count').pipe(my_phi_function).over(whatever))`

Reply:
I’m not opposed to the idea, but I think it still leaves open the question of "how do we handle denominator 0", unless I’m missing something maybe?

Reply:
I'm saying that this logic is complex enough that this part of the chained operators should be broken out into a function

Reply:
I made an issue out of this, since I think we might want to chew on this more and dig deeper as a group?

Reply:
See #69